src/Randomizer.App/MsuGeneratorService.cs

```diff
@@ -78,61 +38,62 @@ public bool EnableMsu1Support(RandomizerOptions options, byte[] rom, string romP
-                ? fallbackCrystalFile
-                : fallbackPendantFile;
-            if (File.Exists(fallbackPcmFile))
-                NativeMethods.CreateHardLink(intendedPcmFile, fallbackPcmFile, IntPtr.Zero);
+            swap = true;
         }
-
-        // Loop through files with fallbacks
-        foreach (var intendedPcmIndex in s_fallbackSongs.Keys.OrderBy(x => x))
+        // Swap if we see don't see the Skull Woods track below 100 and tthat track 133 is the Z3 epilogue music based on if the track does not loop
+        else if (!hasTrack41 && hasTrack133 && !DoesPCMLoop(msuPath.Replace(".msu", "-133.pcm")))
         {
-            var intendedPcmFile = Path.Combine(romFolder!, $"{romBaseName}-{intendedPcmIndex}.pcm");
-            if (File.Exists(intendedPcmFile)) continue;
-            var fallbackPcmIndex = s_fallbackSongs[intendedPcmIndex];
-            var fallbackPcmFile = Path.Combine(romFolder!, $"{romBaseName}-{fallbackPcmIndex}.pcm");
-            if (File.Exists(fallbackPcmFile))
-                NativeMethods.CreateHardLink(intendedPcmFile, fallbackPcmFile, IntPtr.Zero);
+            swap = true;
         }
 
-        // To avoid weirdness with missing boss fanfare, use an empty pcm file in its place
-        var bossVictoryPcmFile = Path.Combine(romFolder!, $"{romBaseName}-119.pcm");
-        if (!File.Exists(bossVictoryPcmFile))
+        // Copy pcm files
+        foreach (var msuFile in Directory.EnumerateFiles(msuFolder!, $"{msuBaseName}*"))
         {
-            using var stream = Assembly.GetExecutingAssembly()
-                .GetManifestResourceStream("Randomizer.App.Resources.empty.pcm");
-            if (stream == null) throw new FileNotFoundException("empty.pcm file not found");
-            using var fileStream = new FileStream(bossVictoryPcmFile, FileMode.Create, FileAccess.Write);
-            stream.CopyTo(fileStream);
+            var fileName = Path.GetFileName(msuFile);
+            var suffix = fileName.Replace(msuBaseName, "");
+            var link = Path.Combine(romFolder!, romBaseName + suffix);
+
+            // Swap SM and Z3 tracks if needed
+            if (swap && suffix.Contains(".pcm"))
+            {
+                if (int.TryParse(suffix.Replace("-", "").Replace(".pcm", ""), out var trackNumber) && trackNumber > 0)
+                {
+                    if (trackNumber < 99)
+                    {
+                        trackNumber += 100;
+                    }
+                    else if (trackNumber > 99)
+                    {
+                        trackNumber -= 100;
+                    }
+
+                    var oldLink = link;
+                    link = Path.Combine(romFolder!, romBaseName + "-" + trackNumber + ".pcm");
+                }
+            }
+            NativeMethods.CreateHardLink(link, msuFile, IntPtr.Zero);
         }
 
         error = "";
         return true;
     }
 
+    private static bool DoesPCMLoop(string fileName)
```

Comment:
This *probably* won't actually be a problem, but some of the tracks in my one MSU pack so far *do* loop on sample zero (including my Metroid track 33). Since it's only BSU packs that we need to worry about, we probably don't need to do anything about this until/unless we find (or somebody reports) an old pack that isn't swapping properly?

Reply:
So I just yoinked the idea of how to detect loops from Skelebro, and my understanding of the PCM format isn't the best. Is the concern that this doesn't detect whether a track loops correctly or not? Is there a better, more proper way of doing it?

That being said, this is very, very much a corner case. This code would only come up if you have an MSU with extended SM tracks and no extended Z3 tracks. The likelihood of that happening is very small, I feel.

But still, if there's a better way of detecting if a PCM file loops or not, it would be good to make the change. It's more relevant on the MSU Randomizer.

Reply:
Yeah, a loop point of zero either means the track should loop the entire file or a loop point wasn't specified when the MSU pack was created. You can't tell for sure from looking at the PCM file. The closest you can get is "definitely does loop" or "may or may not loop".

But yeah, it should be pretty rare that this keeps anpack from getting swapped, as you said. Hopefully if it happens to anybody, they'll report it and we can help them fix their pack.

Reply:
Gotcha. Okay, based on that, I updated it to either check if the Z3 epilogue or SM credits themes are set to loop. If it sees that either are set to loop, then it'll assume a swap is necessary. There are some edge cases, but I think since this is a fallback just in cases where we don't have extended dungeon themes, it's probably fine. The only MSU pack I'm aware of that's like this is the town theme, and this seems to address it.